OrbitProducer/include/OrbitProducer/LockFreeBufferCaptureEventProducer.h

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_PRODUCER_LOCK_FREE_BUFFER_CAPTURE_EVENT_PRODUCER_H_
+#define ORBIT_PRODUCER_LOCK_FREE_BUFFER_CAPTURE_EVENT_PRODUCER_H_
+
+#include "OrbitBase/Logging.h"
+#include "OrbitProducer/CaptureEventProducer.h"
+#include "concurrentqueue.h"
+
+namespace orbit_producer {
+
+// This still abstract implementation of CaptureEventProducer provides a lock-free queue
+// where to write events with low overhead from the fast path where they are produced.
+// The methods EnqueueIntermediateEvent(IfCapturing) allow to enqueue those events.
+//
+// Internally, a thread reads from the lock-free queue and sends CaptureEvents
+// to ProducerSideService using the methods provided by the superclass.
+//
+// Note that the events stored in the lock-free queue, whose type is specified by the
+// type parameter IntermediateEventT, don't need to be CaptureEvents, nor protobufs at all.
+// This is to allow enqueuing objects that are faster to produce than protobufs.
```

Comment:
Whats really the difference here? How can it be faster to produce a non CaptureEvent and let it then translate here to an CaptureEvent then just storing CaptureEvents directly (and creating that on the user side)?
Is it about potential copies of the proto?

Reply:
The difference is that the translation is done by the reader, not by the writer, and the latter is the one that needs to be fast.

Reply:
By the way, I'm not saying that protobufs are so slow, but they do have an overhead (if I understand correctly, mostly because of the memory management, as all fields are optional) which in some cases can be relevant if compared to a simple struct of a few bytes.

Reply:
Okay. Lets see how the users are going to look like. If every implementation will use CaptureEvent anyways, we can remove that later.